% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,10pt,english]{sphinxmanual}
\usepackage[utf8]{inputenc}
\DeclareUnicodeCharacter{00A0}{\nobreakspace}
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage{longtable}
\usepackage{sphinx}
\usepackage{multirow}


\title{sfs\_coder Documentation}
\date{April 08, 2014}
\release{0}
\author{Lawrence Uricchio, Raul Torres, Ryan Hernandez}
\newcommand{\sphinxlogo}{}
\renewcommand{\releasename}{Release}
\makeindex

\makeatletter
\def\PYG@reset{\let\PYG@it=\relax \let\PYG@bf=\relax%
    \let\PYG@ul=\relax \let\PYG@tc=\relax%
    \let\PYG@bc=\relax \let\PYG@ff=\relax}
\def\PYG@tok#1{\csname PYG@tok@#1\endcsname}
\def\PYG@toks#1+{\ifx\relax#1\empty\else%
    \PYG@tok{#1}\expandafter\PYG@toks\fi}
\def\PYG@do#1{\PYG@bc{\PYG@tc{\PYG@ul{%
    \PYG@it{\PYG@bf{\PYG@ff{#1}}}}}}}
\def\PYG#1#2{\PYG@reset\PYG@toks#1+\relax+\PYG@do{#2}}

\expandafter\def\csname PYG@tok@gd\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gu\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@gt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PYG@tok@gs\endcsname{\let\PYG@bf=\textbf}
\expandafter\def\csname PYG@tok@gr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PYG@tok@cm\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@vg\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@m\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@cs\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\colorbox[rgb]{1.00,0.94,0.94}{\strut ##1}}}
\expandafter\def\csname PYG@tok@ge\endcsname{\let\PYG@it=\textit}
\expandafter\def\csname PYG@tok@vc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@il\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@go\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.20,0.20,0.20}{##1}}}
\expandafter\def\csname PYG@tok@cp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@gi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PYG@tok@gh\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PYG@tok@ni\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.84,0.33,0.22}{##1}}}
\expandafter\def\csname PYG@tok@nl\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.13,0.44}{##1}}}
\expandafter\def\csname PYG@tok@nn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@no\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.38,0.68,0.84}{##1}}}
\expandafter\def\csname PYG@tok@na\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@nb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.05,0.52,0.71}{##1}}}
\expandafter\def\csname PYG@tok@nd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.33,0.33,0.33}{##1}}}
\expandafter\def\csname PYG@tok@ne\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@nf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.49}{##1}}}
\expandafter\def\csname PYG@tok@si\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.44,0.63,0.82}{##1}}}
\expandafter\def\csname PYG@tok@s2\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@vi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@nt\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.02,0.16,0.45}{##1}}}
\expandafter\def\csname PYG@tok@nv\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.38,0.84}{##1}}}
\expandafter\def\csname PYG@tok@s1\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@gp\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@sh\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@ow\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@sx\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.78,0.36,0.04}{##1}}}
\expandafter\def\csname PYG@tok@bp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c1\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@kc\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@c\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.50,0.56}{##1}}}
\expandafter\def\csname PYG@tok@mf\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@err\endcsname{\def\PYG@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PYG@tok@kd\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@ss\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.32,0.47,0.09}{##1}}}
\expandafter\def\csname PYG@tok@sr\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.14,0.33,0.53}{##1}}}
\expandafter\def\csname PYG@tok@mo\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@mi\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.13,0.50,0.31}{##1}}}
\expandafter\def\csname PYG@tok@kn\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@o\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PYG@tok@kr\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@s\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@kp\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@w\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PYG@tok@kt\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.56,0.13,0.00}{##1}}}
\expandafter\def\csname PYG@tok@sc\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sb\endcsname{\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@k\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.00,0.44,0.13}{##1}}}
\expandafter\def\csname PYG@tok@se\endcsname{\let\PYG@bf=\textbf\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}
\expandafter\def\csname PYG@tok@sd\endcsname{\let\PYG@it=\textit\def\PYG@tc##1{\textcolor[rgb]{0.25,0.44,0.63}{##1}}}

\def\PYGZbs{\char`\\}
\def\PYGZus{\char`\_}
\def\PYGZob{\char`\{}
\def\PYGZcb{\char`\}}
\def\PYGZca{\char`\^}
\def\PYGZam{\char`\&}
\def\PYGZlt{\char`\<}
\def\PYGZgt{\char`\>}
\def\PYGZsh{\char`\#}
\def\PYGZpc{\char`\%}
\def\PYGZdl{\char`\$}
\def\PYGZhy{\char`\-}
\def\PYGZsq{\char`\'}
\def\PYGZdq{\char`\"}
\def\PYGZti{\char`\~}
% for compatibility with earlier versions
\def\PYGZat{@}
\def\PYGZlb{[}
\def\PYGZrb{]}
\makeatother

\begin{document}

\maketitle
\tableofcontents
\phantomsection\label{index::doc}


Contents:


\chapter{Introduction}
\label{rstfiles/intro:introduction}\label{rstfiles/intro::doc}\label{rstfiles/intro:sfs-coder-documentation}
Forward simulation of DNA sequences is a powerful tool for analyzing the impact
of evolutionary forces on genetic variaiton.  Forward simulation can generate
sequence data under arbitrarily complex models that include nautral selection
as well as complex demography.  These models are difficult to handle
analytically, and other simulation frameworks such as the coalescent cannot
provide the same level of generality.

However, there remain some barriers to the adoption of population genetic
simulators for many members of the genetics community. Simulation software
tools can be daunting to master because of the vast number of input options
and the density of the output data.

sfs\_coder is a python based front end to the popular forward simulation
software SFS\_CODE.  It allows python coders to easily execute and analyze
simulations performed with SFS\_CODE.  For beginning users and coders, it
provides a number of useful cannonical models that are prepackaged and can be
accessed and analyzed with just a few lines of code.

For advanced users, sfs\_coder provides a tool set to access the power of
SFS\_CODE through a python based interface.  We encourage any interested users
to extend the code base that we provide and add any models that could be
useful for the community.

sfs\_coder is free to use and distribute for personal or academic use.
We hope that you will find it useful.


\chapter{Installation}
\label{rstfiles/install:installation}\label{rstfiles/install::doc}
sfs\_coder does not require any installation of its own in order
to get its basic functionality.  Just download the source and put it
wherever you want on your machine. However, some functionality within
sfs\_coder requires external software in order to run.


\section{Required dependencies}
\label{rstfiles/install:required-dependencies}\begin{itemize}
\item {} 
\href{http://sfscode.sourceforge.net}{SFS\_CODE}

\end{itemize}

SFS\_CODE can be installed anywhere on the user's machine. The path
to the binary is supplied to sfs\_coder (see the section on
running SFS\_CODE through sfs\_coder).
\begin{itemize}
\item {} 
\href{https://www.python.org}{python} (\emph{2.7 or greater})

\end{itemize}


\section{Optional dependencies}
\label{rstfiles/install:optional-dependencies}\begin{itemize}
\item {} 
\href{https://code.google.com/p/mpmath/}{mpmath} (\emph{required for rescaled recurrent hitchhiking simulations})

\item {} 
\href{http://www.scipy.org/}{scipy} (\emph{required for rescaled recurrent hitchhiking simulations and some
methods in sfsplot})

\item {} 
\href{http://matplotlib.org/}{matplotlib} (\emph{required for plotting the output})

\end{itemize}


\section{Importing sfs\_coder's modules}
\label{rstfiles/install:importing-sfs-coder-s-modules}
Python uses the PYTHONPATH system variable to search for modules that are
imported.  Suppose we download sfs\_coder and store it in the directory
`\textasciitilde{}/sfs\_coder', and then we try to execute the following script called
`basic.py':

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{command}

\PYG{n}{com} \PYG{o}{=} \PYG{n}{SFSCommand}\PYG{p}{(}\PYG{p}{)}
\end{Verbatim}

If the directory that contains command.py (`\textasciitilde{}/sfs\_coder/src' by default) is not
included in the PYTHONPATH variable, this will result in an error similar to
the following:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{Traceback} \PYG{p}{(}\PYG{n}{most} \PYG{n}{recent} \PYG{n}{call} \PYG{n}{last}\PYG{p}{)}\PYG{p}{:}
  \PYG{n}{File} \PYG{l+s}{\PYGZdq{}}\PYG{l+s}{basic.py}\PYG{l+s}{\PYGZdq{}}\PYG{p}{,} \PYG{n}{line} \PYG{l+m+mi}{13}\PYG{p}{,} \PYG{o+ow}{in} \PYG{o}{\PYGZlt{}}\PYG{n}{module}\PYG{o}{\PYGZgt{}}
    \PYG{k+kn}{import} \PYG{n+nn}{command}
\PYG{n+ne}{ImportError}\PYG{p}{:} \PYG{n}{No} \PYG{n}{module} \PYG{n}{named} \PYG{n}{command}
\end{Verbatim}

Python does not know where the command module is!  To fix this, we can add the
`\textasciitilde{}/sfs\_coder/src' directory to the PYTHONPATH variable in a couple different
ways.


\subsection{Adding the sfs\_coder source directory to PYTHONPATH in .bashrc}
\label{rstfiles/install:adding-the-sfs-coder-source-directory-to-pythonpath-in-bashrc}
If you execute your scripts at the command line with a bash shell, you can add
a line to your .bashrc file that will fix this problem and allow you to run the
above script.  The .bashrc file exists in your home directory and is read by
bash every time you open a new shell.  If the file doesn't exist in your home
directory you can create it.

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n+nb}{export }\PYG{n+nv}{PYTHONPATH}\PYG{o}{=}\PYG{n+nv}{\PYGZdl{}PYTHONPATH}:\PYGZti{}/sfs\PYGZus{}coder/src
\end{Verbatim}

Of course, if the path to your sfs\_coder `src' directory is different than
above you will need to provide the path to your copy of this directory.


\subsection{Adding the path to your sfs\_coder source directory within a python script}
\label{rstfiles/install:adding-the-path-to-your-sfs-coder-source-directory-within-a-python-script}
You can also add the path to sfs\_coder's `src' directory to any python script
if for any reason you don't want to modify your .bashrc as above.  Assuming
the same directory layout as the above example, we can use:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{sys}
\PYG{n}{sys}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{append}\PYG{p}{(}\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{\PYGZti{}/sfs\PYGZus{}coder/src}\PYG{l+s}{\PYGZsq{}}\PYG{p}{)}
\PYG{k+kn}{import} \PYG{n+nn}{command}

\PYG{n}{com} \PYG{o}{=} \PYG{n}{SFSCommand}\PYG{p}{(}\PYG{p}{)}
\end{Verbatim}

This adds `\textasciitilde{}/sfs\_coder/src' to the PYTHONPATH variable within the script.
Note that this solution requires us to add this line of code to every python
script whereas the first solution allows us to import the modules just like
any other python modules.


\chapter{Running SFS\_CODE simulations}
\label{rstfiles/execute::doc}\label{rstfiles/execute:running-sfs-code-simulations}
Running SFS\_CODE simulations with sfs\_coder requires only a few lines
of code.

First, we import the command modile, initialize an SFSCommand object, and tell
the software where the sfs\_code binary is located.

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{command}

\PYG{n}{com} \PYG{o}{=} \PYG{n}{SFSCommand}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{com}\PYG{o}{.}\PYG{n}{sfs\PYGZus{}code\PYGZus{}loc} \PYG{o}{=} \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{/path/to/sfs\PYGZus{}code}\PYG{l+s}{\PYGZsq{}}
\end{Verbatim}

Next, we need to build a command line.  Although this process is very
flexible (in fact we can build any command line that is accepted by SFS\_CODE),
we have prepackaged several models that may be of general interest.  For
example, to simulate the model of Gutenkunst (2009, \emph{PLoS Genetics}), we
call the following:

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{n}{com}\PYG{o}{.}\PYG{n}{gutenkunst}\PYG{p}{(}\PYG{p}{)}

\PYG{n}{com}\PYG{o}{.}\PYG{n}{execute}\PYG{p}{(}\PYG{p}{)}
\end{Verbatim}

And that's it!  Of course, there are many more options that can be added to
modify the parameters of the simulation.  Please see the ``scripts'' directory
in the top level of sfs\_coder for more complicated examples. Below, we include
a slight modification of the above script that demonstrates some basic
functionality that may be useful to users.

\begin{Verbatim}[commandchars=\\\{\}]
\PYG{k+kn}{import} \PYG{n+nn}{command}
\PYG{k+kn}{import} \PYG{n+nn}{os}
\PYG{k+kn}{from} \PYG{n+nn}{random} \PYG{k+kn}{import} \PYG{n}{randint}

\PYG{c}{\PYGZsh{} initialize an SFS\PYGZus{}CODE command, set the prefix of the output subdirectory}
\PYG{n}{com} \PYG{o}{=} \PYG{n}{command}\PYG{o}{.}\PYG{n}{SFSCommand}\PYG{p}{(}\PYG{n}{prefix}\PYG{o}{=}\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{guten.N500}\PYG{l+s}{\PYGZsq{}}\PYG{p}{)}

\PYG{c}{\PYGZsh{} build the command line for the gutenkunst model with N=500, etc}
\PYG{n}{com}\PYG{o}{.}\PYG{n}{gutenkunst}\PYG{p}{(}\PYG{n}{N}\PYG{o}{=}\PYG{l+m+mi}{500}\PYG{p}{,}\PYG{n}{nsam}\PYG{o}{=}\PYG{l+m+mi}{50}\PYG{p}{,}\PYG{n}{nsim}\PYG{o}{=}\PYG{l+m+mi}{10}\PYG{p}{)}

\PYG{c}{\PYGZsh{} set the location of sfs\PYGZus{}code, set the prefix of the out files}
\PYG{n}{com}\PYG{o}{.}\PYG{n}{sfs\PYGZus{}code\PYGZus{}loc} \PYG{o}{=} \PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{join}\PYG{p}{(}\PYG{n}{os}\PYG{o}{.}\PYG{n}{path}\PYG{o}{.}\PYG{n}{expanduser}\PYG{p}{(}\PYG{l+s}{\PYGZsq{}}\PYG{l+s}{\PYGZti{}}\PYG{l+s}{\PYGZsq{}}\PYG{p}{)}\PYG{p}{,}
                       \PYG{l+s}{\PYGZsq{}}\PYG{l+s}{path/to/sfs\PYGZus{}code}\PYG{l+s}{\PYGZsq{}}\PYG{p}{)}

\PYG{c}{\PYGZsh{} execute the command, supplying a random number}
\PYG{n}{com}\PYG{o}{.}\PYG{n}{execute}\PYG{p}{(}\PYG{n}{rand}\PYG{o}{=}\PYG{n}{randint}\PYG{p}{(}\PYG{l+m+mi}{1}\PYG{p}{,}\PYG{l+m+mi}{100000}\PYG{p}{)}\PYG{p}{)}
\end{Verbatim}


\chapter{Analyzing the output}
\label{rstfiles/analyze:analyzing-the-output}\label{rstfiles/analyze::doc}

\chapter{Plotting}
\label{rstfiles/plot::doc}\label{rstfiles/plot:plotting}\phantomsection\label{index:module-command}\index{command (module)}\index{Command (class in command)}

\begin{fulllineitems}
\phantomsection\label{index:command.Command}\pysigline{\strong{class }\code{command.}\bfcode{Command}}
This class stores information from parsed command lines
and provides the mechanics to call SFS\_CODE/ms commands.
\index{add\_out() (command.Command method)}

\begin{fulllineitems}
\phantomsection\label{index:command.Command.add_out}\pysiglinewithargsret{\bfcode{add\_out}}{}{}
Adds the path to an output directory to self.line.  This method takes
no arguments but uses self.prefix (a label for the set of simulation
experiments you are doing) and self.outdir (a label for the directory
in which sets of simulations are stored).

Thus, the full path is set to:
\begin{itemize}
\item {} 
\emph{os.path.join(self.outdir,self.prefix)}

\end{itemize}

The values of self.outdir and self.prefix are set when an object is
instantiated.  See the documentation of command.SFSCommand() for the 
defaults of these values.

\end{fulllineitems}

\index{execute() (command.Command method)}

\begin{fulllineitems}
\phantomsection\label{index:command.Command.execute}\pysiglinewithargsret{\bfcode{execute}}{\emph{rand=1}}{}
execute a simulation command
\begin{itemize}
\item {} 
Parameters:
\begin{itemize}
\item {} \begin{description}
\item[{\emph{rand=1} }] \leavevmode
a random integer. If the value is not reset by the user
then a new random number is rolled within self.execute.

\end{description}

\end{itemize}

\end{itemize}

\end{fulllineitems}


\end{fulllineitems}

\index{SFSCommand (class in command)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand}\pysiglinewithargsret{\strong{class }\code{command.}\bfcode{SFSCommand}}{\emph{outdir='/Users/luricchio/projects/cluster\_backup/sfs\_coder/doc/sims'}, \emph{prefix='out'}, \emph{err='err'}}{}
This class is used to store, parse, and convert SFS\_CODE command lines.

Upon initialization, an object of the SFSCommand class sets the values
of many of its attributes to the SFS\_CODE defaults.
\begin{itemize}
\item {} 
Parameters:
\begin{itemize}
\item {} \begin{description}
\item[{\emph{outdir=os.path.join(os.getcwd(), `sims')} }] \leavevmode
A directory containing
subdirectories with sfs\_code simulations.

\end{description}

\item {} \begin{description}
\item[{\emph{prefix='out'} }] \leavevmode
The prefix of the out directory and the data files 
within the out directory.

\end{description}

\item {} \begin{description}
\item[{\emph{err='err'}}] \leavevmode
The name of the directory that contains all the stderr ouput from
calling sfs\_code.

\end{description}

\end{itemize}

\item {} 
Attributes:
\begin{itemize}
\item {} \begin{description}
\item[{\emph{self.com\_string='`}}] \leavevmode
the entire command stored as a single string.

\end{description}

\item {} \begin{description}
\item[{\emph{self.outdir= outdir}   }] \leavevmode
the parent directory of output directories for sets of SFS\_CODE 
simulations

\end{description}

\item {} \begin{description}
\item[{\emph{self.sfs\_code\_loc = `'}   }] \leavevmode
the location of the SFS\_CODE binary.

\end{description}

\item {} \begin{description}
\item[{\emph{self.N = 500}  }] \leavevmode
the number of individuals in the ancestral population.

\end{description}

\item {} \begin{description}
\item[{\emph{self.P = {[}2{]}}  }] \leavevmode
the ploidy of the individuals in each population.

\end{description}

\item {} \begin{description}
\item[{\emph{self.t = 0.001}}] \leavevmode
\(\theta = 4Nu = 0.001\).  This is the value of 
\(\theta\) in the ancestral population

\end{description}

\item {} \begin{description}
\item[{\emph{self.L = {[}5000{]}} }] \leavevmode
an array containing the length of each simulate locus.

\end{description}

\item {} \begin{description}
\item[{\emph{self.B = 5 self.p{[}0{]} self.N} }] \leavevmode
the length of the burn in (generations).

\end{description}

\item {} \begin{description}
\item[{\emph{self.prefix= prefix} }] \leavevmode
the prefix for the output file directory and each simulation file.

\end{description}

\item {} \begin{description}
\item[{\emph{self.r=0.0.} }] \leavevmode
\(\rho = 4Nr = 0.0\).  The value of \(\rho\) in the 
ancestral population.

\end{description}

\item {} \begin{description}
\item[{\emph{self.n\_pops=1} }] \leavevmode
number of populations.

\end{description}

\item {} \begin{description}
\item[{\emph{self.n\_iter=1} }] \leavevmode
number of simulations.

\end{description}

\item {} \begin{description}
\item[{\emph{self.line={[}{]}} }] \leavevmode
an array of strings, each of which is an argument
to SFS\_CODE.  This is the attribute that is used to execute SFS\_CODE
commands.

\end{description}

\end{itemize}

\end{itemize}
\index{add\_rand() (command.SFSCommand method)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand.add_rand}\pysiglinewithargsret{\bfcode{add\_rand}}{\emph{rand}}{}
Just a little helper method to add random numbers to command lines.
Not really recommended for use outside of this module.

\end{fulllineitems}

\index{build\_BGS() (command.SFSCommand method)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand.build_BGS}\pysiglinewithargsret{\bfcode{build\_BGS}}{\emph{n\_sim=10}, \emph{theta=0.0001}, \emph{recomb=False}, \emph{rho=0.001}, \emph{N=250}, \emph{n\_sam=10}, \emph{alpha=5}, \emph{L=100000}, \emph{Lmid=100000}}{}
A method for running simulations of background selection.  This method
is a bit underdeveloped at the moment, so stay tuned for more on how
it works.

\end{fulllineitems}

\index{build\_RHH() (command.SFSCommand method)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand.build_RHH}\pysiglinewithargsret{\bfcode{build\_RHH}}{\emph{alpha=1000.0}, \emph{N0=5000.0}, \emph{rho0=0.001}, \emph{lam0=1e-10}, \emph{delta=0.01}, \emph{L0=-1}, \emph{L1=100000.0}, \emph{loop\_max=10}, \emph{L\_neut=1000.0}, \emph{theta\_neut=0.001}, \emph{minpop=100}, \emph{recomb\_dir='./recombfiles'}, \emph{outdir='sims'}, \emph{TE=2}, \emph{r\_within=False}, \emph{neg\_sel\_rate=0.0}, \emph{alpha\_neg=5}, \emph{additive=1}, \emph{Lextend=1}, \emph{mutation=}\optional{}, \emph{bottle=}\optional{}, \emph{expansion=}\optional{}}{}
A method to build a recurrent hitchhiking command line using the 
method of Uricchio \& Hernandez (2014, \emph{Genetics}).
\begin{itemize}
\item {} 
Dependencies:
\begin{itemize}
\item {} 
scipy

\item {} 
mpmath

\end{itemize}

\item {} 
Parameters
\begin{itemize}
\item {} \begin{description}
\item[{\emph{alpha = 1000} }] \leavevmode
\(\alpha = 2Ns\), the ancestral population 
scaled strength of selection.  Note that demographic events 
can change N, and hence they also changle alpha.

\end{description}

\item {} \begin{description}
\item[{\emph{N0 = 5000} }] \leavevmode
the ancestral population size

\end{description}

\item {} \begin{description}
\item[{\emph{rho0 = 0.001} }] \leavevmode
the population scaled recombination coefficient
in the ancestral population.

\end{description}

\item {} \begin{description}
\item[{\emph{lam0 = 10*}-10 }] \leavevmode
the rate of positive substitutions per generation
per site in the population.

\end{description}

\item {} \begin{description}
\item[{\emph{delta = 0.01} }] \leavevmode
a single parameter that encapsulates both delta 
parameters from Uricchio \& Hernandez (Genetics, 2014).
Smaller values of delta result in dynamics that are a better
match for the original population of size N0, but are more 
computationally expensive.  We do not recommend using values
of delta greater than 0.1.  For more information please see the
paper referenced above.

\end{description}

\item {} \begin{description}
\item[{\emph{L0 = -1} }] \leavevmode
the length of the flanking sequence on each side of the
neutral locus. If L0 is not reset from it's default value, 
it is automatically set to L0 = s0/r0, where s0 and r0 are 
alpha/2N0 and rho/4N0, respectively.

\end{description}

\item {} \begin{description}
\item[{\emph{theta\_neut = 0.001} }] \leavevmode
the neutral value of theta.

\end{description}

\item {} \begin{description}
\item[{\emph{TE=2}}] \leavevmode
the ending time of the simulation in units of 
2*N0*self.P{[}0{]} generations.

\end{description}

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{build\_genomic() (command.SFSCommand method)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand.build_genomic}\pysiglinewithargsret{\bfcode{build\_genomic}}{\emph{basedir='/Users/luricchio/projects/cluster\_backup/sfs\_coder/doc/req'}, \emph{outdir='/Users/luricchio/projects/cluster\_backup/sfs\_coder/doc/input\_files'}, \emph{datafile='hg19\_gencode.v14.gtf.gz'}, \emph{chr=2}, \emph{begpos=134545415}, \emph{endpos=138594750}, \emph{db=136545415}, \emph{de=136594750}, \emph{withseq=0}, \emph{fafile='hg19.fa.gz'}, \emph{phast\_file='hg19\_phastCons\_mammal.wig'}, \emph{dense\_dist=5000}, \emph{regname='`}, \emph{annotfile='`}, \emph{recombout='`}, \emph{sel=True}}{}
A method to build simulations of genomic regions with realistic 
genome structure.  The default data sources and options are all
human-centric, but in principle these methods could be used to simulate
sequences from any population for which the relevant data sources are 
available.

\end{fulllineitems}

\index{convert\_sfs\_ms() (command.SFSCommand method)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand.convert_sfs_ms}\pysiglinewithargsret{\bfcode{convert\_sfs\_ms}}{}{}
A method to convert an sfs\_code command line to ms.  Most but not all
switches are converted (the method should let you know if you try to
use an unsupported switch).  The method will ignore selection 
based switches in SFS\_CODE (since ms only includes neutral 
demographic models).

This method has been tested only sparingly and it is recommended that 
it is used with great caution.

\end{fulllineitems}

\index{genomic() (command.SFSCommand method)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand.genomic}\pysiglinewithargsret{\bfcode{genomic}}{\emph{basedir='/Users/luricchio/projects/cluster\_backup/sfs\_coder/doc/req'}, \emph{outdir='/Users/luricchio/projects/cluster\_backup/sfs\_coder/doc/input\_files'}, \emph{datafile='hg19\_gencode.v14.gtf.gz'}, \emph{chr=2}, \emph{begpos=134545415}, \emph{endpos=138594750}, \emph{db=136545415}, \emph{de=136594750}, \emph{withseq=0}, \emph{fafile='hg19.fa.gz'}, \emph{phast\_file='hg19\_phastCons\_mammal.wig'}, \emph{dense\_dist=5000}, \emph{N=2000}, \emph{mutation=}\optional{}, \emph{sel=True}}{}
A method for running simulations of genomic elements using realistic
genome structure. Most of the heavy lifting is done by the 
build\_genomic() method.  This method exists mostly as a slimmed down
interface to that method, with the major difference being that this 
version allows for the inclusion of a demographic model. Currenrly,
only the demographic model of Gutenkunst (2009, \emph{PLoS Genetics}) is
included.

\end{fulllineitems}

\index{gutenkunst() (command.SFSCommand method)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand.gutenkunst}\pysiglinewithargsret{\bfcode{gutenkunst}}{\emph{add\_on=False}, \emph{nsim=1}, \emph{N=10000}, \emph{non\_coding=False}, \emph{recombfile='`}, \emph{nsam=100}, \emph{mutation=}\optional{}, \emph{loci=}\optional{}, \emph{sel=}\optional{}, \emph{L=}\optional{}}{}
A method that adds the Gutenkunst (2009, \emph{PLoS Genetics}) model to
an SFS\_CODE command line.

\end{fulllineitems}

\index{parse\_string() (command.SFSCommand method)}

\begin{fulllineitems}
\phantomsection\label{index:command.SFSCommand.parse_string}\pysiglinewithargsret{\bfcode{parse\_string}}{}{}
A method to parse SFS\_CODE command lines.  By default, every switch
is stored as an array with the exception of certain special cases that
are stored as dictionaries.

Note, \textbf{only the short form of SFS\_CODE options are currently fully 
supported!}  For example, \emph{-t 0.002} is supported but \emph{--theta 0.002}
is not.

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{index:module-sfs}\index{sfs (module)}\index{Simulation (class in sfs)}

\begin{fulllineitems}
\phantomsection\label{index:sfs.Simulation}\pysigline{\strong{class }\code{sfs.}\bfcode{Simulation}}
A class to store the data from SFS\_CODE simulations.
\index{calc\_S() (sfs.Simulation method)}

\begin{fulllineitems}
\phantomsection\label{index:sfs.Simulation.calc_S}\pysiglinewithargsret{\bfcode{calc\_S}}{\emph{pop=0}, \emph{multi\_skip=True}, \emph{loci=}\optional{}}{}
calculate the number of segretating sites within a specific
population.
\begin{itemize}
\item {} 
Parameters:
\begin{itemize}
\item {} 
\emph{pop = 0}
population in which to calculate the number of segregating sites.

\item {} 
\emph{multi\_skip = True}
skip sites that are more than biallelic if true

\item {} 
\emph{loci = {[}{]}}
A list of loci over which to calculate the number of segregating
sites.  Uses all loci if this is left blank.

\end{itemize}

\end{itemize}

\end{fulllineitems}

\index{make\_muts() (sfs.Simulation method)}

\begin{fulllineitems}
\phantomsection\label{index:sfs.Simulation.make_muts}\pysiglinewithargsret{\bfcode{make\_muts}}{}{}
This method takes the string of data that is generated by readsfs and 
parses it into individual mutations.  It handles some of the oddities
of how mutations are stored when they segregate in multiple 
populations.

This is mostly intended for internal use within this module, so use 
or modify at your own risk.

\end{fulllineitems}


\end{fulllineitems}

\phantomsection\label{index:module-sfsplot}\index{sfsplot (module)}\phantomsection\label{index:module-ms}\index{ms (module)}\phantomsection\label{index:module-readsfs}\index{readsfs (module)}\index{FileData (class in readsfs)}

\begin{fulllineitems}
\phantomsection\label{index:readsfs.FileData}\pysiglinewithargsret{\strong{class }\code{readsfs.}\bfcode{FileData}}{\emph{file='`}}{}
A class that handles the basic parsing of sfs\_code output file
data.
\begin{itemize}
\item {} 
Parameters:
\begin{itemize}
\item {} 
\emph{file= `'}
\begin{quote}

the path to the file that is to be read.
\end{quote}

\end{itemize}

\item {} 
Attributes:
\begin{itemize}
\item {} \begin{description}
\item[{\emph{self.file = file}}] \leavevmode
the path to the file that is to be read.

\end{description}

\item {} \begin{description}
\item[{\emph{sims = {[}{]}}}] \leavevmode
an array of sfs.Simulation objects.

\end{description}

\end{itemize}

\end{itemize}
\index{get\_sims() (readsfs.FileData method)}

\begin{fulllineitems}
\phantomsection\label{index:readsfs.FileData.get_sims}\pysiglinewithargsret{\bfcode{get\_sims}}{}{}
A method that reads sfs\_code output files and stores all the data in
sfs.Simulation objects.

\end{fulllineitems}


\end{fulllineitems}

\index{msData (class in readsfs)}

\begin{fulllineitems}
\phantomsection\label{index:readsfs.msData}\pysiglinewithargsret{\strong{class }\code{readsfs.}\bfcode{msData}}{\emph{file='`}}{}
A class to store ms output file data.
\index{get\_sims() (readsfs.msData method)}

\begin{fulllineitems}
\phantomsection\label{index:readsfs.msData.get_sims}\pysiglinewithargsret{\bfcode{get\_sims}}{}{}
A method that parses ms output files and stores each simulation
as an ms.Simulation object.  This method is \emph{not} fully developed
at this time and currently does not store several essential pieces of
information, such as the positions of the segregating sites.

\end{fulllineitems}


\end{fulllineitems}



\chapter{Indices and tables}
\label{index:indices-and-tables}\begin{itemize}
\item {} 
\emph{genindex}

\item {} 
\emph{modindex}

\item {} 
\emph{search}

\end{itemize}


\renewcommand{\indexname}{Python Module Index}
\begin{theindex}
\def\bigletter#1{{\Large\sffamily#1}\nopagebreak\vspace{1mm}}
\bigletter{c}
\item {\texttt{command}}, \pageref{index:module-command}
\indexspace
\bigletter{m}
\item {\texttt{ms}}, \pageref{index:module-ms}
\indexspace
\bigletter{r}
\item {\texttt{readsfs}}, \pageref{index:module-readsfs}
\indexspace
\bigletter{s}
\item {\texttt{sfs}}, \pageref{index:module-sfs}
\item {\texttt{sfsplot}}, \pageref{index:module-sfsplot}
\end{theindex}

\renewcommand{\indexname}{Index}
\printindex
\end{document}
